cli/pcluster/utils.py

```diff
@@ -1225,25 +1170,122 @@ def get_ebs_snapshot_info(ebs_snapshot_id, raise_exceptions=False):
         )
 
 
-def get_instance_network_interfaces(instance_type, instance_info=None):
-    """Return the number of network interfaces to configure for the instance type."""
-    if not instance_info:
-        instance_info = get_instance_type(instance_type)
+class Cache:
+    """Simple utility class providing a cache mechanism for expensive functions."""
+
+    caches = []
+
+    @staticmethod
+    def enabled():
+        """Tell if the cache is enabled."""
+        return not os.environ.get("PCLUSTER_CACHE_DISABLED")
+
+    @staticmethod
+    def clear_all():
+        """Clear the content of all caches."""
+        for cache in Cache.caches:
+            cache.clear()
+
+    @staticmethod
+    def _make_key(args, kwargs):
+        key = args
+        if kwargs:
+            for item in kwargs.items():
+                key += item
+        return hash(key)
+
+    @staticmethod
+    def cached(function):
+        """
+        Decorate a function to make it use a results cache based on passed arguments.
+
+        Note: all arguments must be hashable for this function to work properly.
+        """
+        cache = {}
+        Cache.caches.append(cache)
+
+        @functools.wraps(function)
+        def wrapper(*args, **kwargs):
+            cache_key = Cache._make_key(args, kwargs)
 
-    # Until maximumNetworkCards is not available, 1 is a safe value for all instance types
-    needed_interfaces = int(instance_info.get("NetworkInfo").get("MaximumNetworkCards", 1))
+            if Cache.enabled() and cache_key in cache:
+                return cache[cache_key]
+            else:
+                rv = function(*args, **kwargs)
+                if Cache.enabled() and rv:
+                    cache[cache_key] = rv
+                return rv
 
-    return needed_interfaces
+        return wrapper
 
 
-def get_instance_gpus(instance_type, instance_info=None):
-    """Return the number of GPUs provided by the instance type."""
-    if not instance_info:
-        instance_info = get_instance_type(instance_type)
+class InstanceTypeInfo:
+    """Data object wrapping the result of a describe_instance_types call."""
+
+    def __init__(self, instance_type_data):
+        self.instance_type_data = instance_type_data
+
+    @staticmethod
+    @Cache.cached
+    def init_from_instance_type(instance_type, exit_on_error=True):
```

Comment:
I would not generalize it to init cause it would be harder to add multiple factories if needed. I can try to pick a better name if that is not good enough.